### PR TITLE
[Binary parser] Add dedup processing and parser-owned example object pool for dedup

### PR DESCRIPTION
**Pull request by olgavrou**
De-dedup but no decompression yet

To be merged after: https://github.com/VowpalWabbit/vowpal_wabbit/pull/2848 which implements dedup parsing and object re-use on the json parser

### CODE REVIEW

File: external_parser/example_joiner.cc [hunk 1]
```diff
@@ -217,11 +268,58 @@ int example_joiner::process_outcome(const v2::Event &event,
...
+  auto dedup = v2::GetDedupInfo(event.payload()->data());
+
+  for (size_t i = 0; i < dedup->ids()->size(); i++) {
+
+    auto examples = v_init<example *>();
+    examples.push_back(get_or_create_example());
+
+    if (_vw->audit || _vw->hash_inv) {
+      VW::template read_line_json<true>(
+          *_vw, examples, const_cast<char *>(dedup->values()->Get(i)->c_str()),
```
Comment (kumpera) [inline external_parser/example_joiner.cc:296]: This need some error checking as all fields are optional.

File: external_parser/example_joiner.cc [hunk 2]
```diff
@@ -217,11 +268,58 @@ int example_joiner::process_outcome(const v2::Event &event,
   return 0;
 }
 
+int example_joiner::process_dedup(const v2::Event &event,
+                                  const v2::Metadata &metadata) {
+
+  // new dedup payload, we need to clear the old one
```
Comment (olgavrou) [inline external_parser/example_joiner.cc:274]: intra-batch sharing in upcoming PR

**Review by kumpera — APPROVED** [review]

File: external_parser/example_joiner.cc [hunk 3]
```diff
@@ -217,11 +268,58 @@ int example_joiner::process_outcome(const v2::Event &event,
...
+      VW::template read_line_json<true>(
+          *_vw, examples, const_cast<char *>(dedup->values()->Get(i)->c_str()),
+          get_or_create_example_f, this);
+    } else {
+      VW::template read_line_json<false>(
+          *_vw, examples, const_cast<char *>(dedup->values()->Get(i)->c_str()),
+          get_or_create_example_f, this);
+    }
+
+    _dedup_examples.emplace(dedup->ids()->Get(i), examples[0]);
```
Comment (cheng-tan) [inline external_parser/example_joiner.cc:304]: also needs error handling if id exists.

File: external_parser/example_joiner.cc [hunk 4]
```diff
@@ -72,9 +74,50 @@ float earliest(const joined_event &event) { return 0.0; }
...
 example_joiner::example_joiner(vw *vw) : _vw(vw) {}
 
+example_joiner::~example_joiner() {
+  // cleanup examples
+  for (auto *ex : _example_pool) {
+    VW::dealloc_examples(ex, 1);
+  }
+}
+
+example *example_joiner::get_or_create_example() {
```
[the remark this answers is not in the document]
Comment (olgavrou) [reply external_parser/example_joiner.cc:84]: Yes this is copy/pasted from safe_vw, however external_builder and rlclientlib don't share any code atm and are built independently so it won't be that easy to pull that code out for re-use

File: external_parser/example_joiner.cc (hunk 3 above)
Comment (olgavrou) [reply external_parser/example_joiner.cc:304]: I'll add TODOs

**Review by cheng-tan — APPROVED** [review]

**Pull request merged**